querydsl-sql/src/main/java/com/querydsl/sql/SQLServerTemplates.java

```diff
@@ -151,7 +151,7 @@ protected SQLServerTemplates(Set<String> keywords, char escape, boolean quote) {
         // truncates timestamps by replacing suffix
         add(Ops.DateTimeOps.TRUNC_YEAR,   "CONVERT(DATETIME, CONVERT(VARCHAR(4), {0}, 120) + '-01-01')");
         add(Ops.DateTimeOps.TRUNC_MONTH,  "CONVERT(DATETIME, CONVERT(VARCHAR(7), {0}, 120) + '-01')");
-        // TODO week
+        add(Ops.DateTimeOps.TRUNC_WEEK, "cast(floor(cast(dateadd(dd, 2 - datepart(dw, {0}), {0}) as float)) as datetime");
```

Comment:
It doesn't follow existing pattern because it wasn't possible to implement it that way. New implementation makes use of internal `date` to `float` conversion, which is not the ideal thing, but is not likely to change. With proper integration test coverage it is better than no implementation at all (I hope).

Reply:
It's better than nothing 🤷🏻 

At [Blaze-Persistence](https://github.com/Blazebit/blaze-persistence/blob/master/core/impl/src/main/java/com/blazebit/persistence/impl/function/trunc/week/MSSQLTruncWeekFunction.java) we use `DATEADD(WEEK, DATEDIFF(WEEK, 0, ?1), 0)` from https://stackoverflow.com/a/32955740/2104280 for SQL Server. Perhaps a bit cleaner 😄

Reply:
Thank you for good suggestion. I've played with it and it seems not to work on some inputs

```
set datefirst 1;
select DATEADD(WEEK, DATEDIFF(WEEK, 0, cast('2021-02-21' as datetime)), 0);
==> 2021-02-22 00:00:00.000
```

I've adopted a similar approach which seems to work correctly and always truncates to Monday.

```
select DATEADD(WEEK, DATEDIFF(WEEK, 0, {0} - 1), 0);
```

The only difference - we subtract one day from the input. There is an [article](http://www.silota.com/docs/recipes/sql-server-date-parts-truncation.html) describing why this works. It not trivial, but it worked on my tests (on different `datefirst` values it always truncates to Monday).